### PR TITLE
Run CoreCLR runtime tests using Mono with LLVM AOT.

### PR DESCRIPTION
**Pull request by imhameed**
- coreclr/build-test.sh now has a new subcommand: 'mono_aot', which builds a
new target, named 'MonoAotCompileTests', added to
coreclr/tests/src/runtest.proj. This target compiles the runtime tests using
Mono LLVM AOT in a simple configuration where the host platform is identical to
the target platform. Parallel compilation happens via a hack: actual
compilation happens in mono/msbuild/aot-compile.proj, a single-target msbuild
file, and runtest.proj invokes this single-purpose project and target using
batching to create multiple parallelizable instances of this project. Future
work: use the MonoAOTCompiler custom task currently used to build the iOS
sample program.

- Avoid using the runtimeVariant string when defining
coreClrProductArtifactName in mono/templates/xplat-pipeline-job.yml. There are
no "runtime variants" of CoreCLR configured with this parameter; instead,
depend on a shared non-runtime-variant build of CoreCLR.

- Mark function DISubprograms as local definitions--this is an LLVM 9
compatibility fix.

- Use --tag=CXX when linking libmonosgen-2.0.so via libtool when LLVM is linked
into Mono.

  This makes libtool use the C++ compiler driver when linking Mono--which uses
whatever platform-specific flags are necessary to link against the C++ stdlib.
Previously, libtool would use the C compiler driver, which didn't do this and
would produce shared objects with no explicit dependency on libstdc++.

  This problem is normally masked because of the very lax dynamic linking
semantics on ELF, but Mono on our CI setup is built in a CentOS 7 image (which
does not contain a C++11 libstdc++) that has a GCC 7 compatibility package
installed, along with a clang 9 installation that detects headers from the GCC
7 compatibility package. This compatibility package includes a libstdc++ linker
script that links C++11 libstdc++ components statically into the target while
dynamically linking against components present in pre-C++11 libstdc++. The end
result of all of this is that Mono built with this configuration will
dynamically depend on C++11 libstdc++ symbols that should have been statically
linked into the library, and will outright fail to run on machines without a
newer version of libstdc++ available.

- Add tests that fail after LLVM AOT compilation to issues.targets.

### CODE REVIEW

**Comment by imhameed** [comment]
AOT LLVM compilation currently emits several messages of the form:
```
2020-07-07T04:34:20.4888844Z   scope points into the type hierarchy
2020-07-07T04:34:20.4893639Z   !22 = !DILocation(line: 0, scope: !19)
2020-07-07T04:34:20.4894676Z   scope points into the type hierarchy
2020-07-07T04:34:20.4895405Z   !23 = !DILocation(line: 21, column: 9, scope: !19)
2020-07-07T04:34:20.4981897Z   scope points into the type hierarchy
2020-07-07T04:34:20.4983023Z   !25 = !DILocation(line: 22, column: 9, scope: !19)
2020-07-07T04:34:20.4989425Z   scope points into the type hierarchy
2020-07-07T04:34:20.4990415Z   !27 = !DILocation(line: 25, column: 13, scope: !19)
```

I tried this out locally against a copy of LLVM with assertions enabled. We're failing here:
```
corerun: /home/imran/ms/mono/llvm-9.x/src/llvm/lib/IR/Metadata.cpp:634: void llvm::MDNode::resolveCycles(): Assertion `!N->isTemporary() && "Expected all forward declarations to be resolved"' failed.
...
Thread 1 (Thread 0x7fcdc216b740 (LWP 128841)):
#0  0x00007fcdc1b57b5a in __waitpid (pid=128844, stat_loc=0x7fff2e4b517c, options=0) at ../sysdeps/unix/sysv/linux/waitpid.c:29
#1  0x00007fcdbe4aa776 in dump_native_stacktrace (signal=<optimized out>, mctx=<optimized out>) at /home/imran/ms/mono/dnr-90/runtime/src/mono/mono/mini/mini-posix.c:1061
#2  mono_dump_native_crash_info (signal=<optimized out>, mctx=<optimized out>, info=<optimized out>) at /home/imran/ms/mono/dnr-90/runtime/src/mono/mono/mini/mini-posix.c:1105
#3  0x00007fcdbe422aa5 in mono_handle_native_crash (signal=0x7fcdbfb30542 "SIGABRT", mctx=0x7fff2e4b5c90, info=0x7fff2e4b5f70, context=0x7fff2e4b5e40) at /home/imran/ms/mono/dnr-90/runtime/src/mono/mono/mini/mini-exceptions.c:3449
#4  0x00007fcdbe4a9bd8 in sigabrt_signal_handler (_dummy=6, _info=0x7fff2e4b5f70, context=0x7fff2e4b5e40) at /home/imran/ms/mono/dnr-90/runtime/src/mono/mono/mini/mini-posix.c:235
#5  <signal handler called>
#6  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:51
#7  0x00007fcdc0f3f42a in __GI_abort () at abort.c:89
#8  0x00007fcdc0f36e67 in __assert_fail_base (fmt=<optimized out>, assertion=assertion@entry=0x7fcdc07aac08 "!N->isTemporary() && \"Expected all forward declarations to be resolved\"", file=file@entry=0x7fcdc07aa968 "/home/imran/ms/mono/llvm-9.x/src/llvm/lib/IR/Metadata.cpp", line=line@entry=634, function=function@entry=0x7fcdc07c4920 <llvm::MDNode::resolveCycles()::__PRETTY_FUNCTION__> "void llvm::MDNode::resolveCycles()") at assert.c:92
#9  0x00007fcdc0f36f12 in __GI___assert_fail (assertion=0x7fcdc07aac08 "!N->isTemporary() && \"Expected all forward declarations to be resolved\"", file=0x7fcdc07aa968 "/home/imran/ms/mono/llvm-9.x/src/llvm/lib/IR/Metadata.cpp", line=634, function=0x7fcdc07c4920 <llvm::MDNode::resolveCycles()::__PRETTY_FUNCTION__> "void llvm::MDNode::resolveCycles()") at assert.c:101
#10 0x00007fcdbf9a1ec1 in llvm::MDNode::resolveCycles() () from /home/imran/ms/mono/dnr-90/runtime/artifacts/tests/coreclr/Linux.x64.Release/Tests/Core_Root/libcoreclr.so
#11 0x00007fcdbf8f5372 in llvm::DIBuilder::finalize() () from /home/imran/ms/mono/dnr-90/runtime/artifacts/tests/coreclr/Linux.x64.Release/Tests/Core_Root/libcoreclr.so
#12 0x00007fcdbe468702 in emit_dbg_info (module=<optimized out>, filename=<optimized out>, cu_name=<optimized out>) at /home/imran/ms/mono/dnr-90/runtime/src/mono/mono/mini/mini-llvm.c:11331
#13 mono_llvm_emit_aot_module (filename=0x55e25c45a6a0 "mono_aot_114vaS/temp.bc", cu_name=<optimized out>) at /home/imran/ms/mono/dnr-90/runtime/src/mono/mono/mini/mini-llvm.c:11257
#14 0x00007fcdbe3faf6c in emit_llvm_file (acfg=<optimized out>) at /home/imran/ms/mono/dnr-90/runtime/src/mono/mono/mini/aot-compiler.c:9940
#15 emit_aot_image (acfg=<optimized out>) at /home/imran/ms/mono/dnr-90/runtime/src/mono/mono/mini/aot-compiler.c:14263
#16 0x00007fcdbe3f2a1c in mono_compile_assembly (ass=0x55e25c3d2970, opts=<optimized out>, aot_options=<optimized out>, global_aot_state=0x7fff2e4b7a50) at /home/imran/ms/mono/dnr-90/runtime/src/mono/mono/mini/aot-compiler.c:13983
#17 0x00007fcdbe3e612d in main_thread_handler (user_data=<optimized out>) at /home/imran/ms/mono/dnr-90/runtime/src/mono/mono/mini/driver.c:1446
#18 mono_main (argc=<optimized out>, argv=<optimized out>) at /home/imran/ms/mono/dnr-90/runtime/src/mono/mono/mini/driver.c:2754
#19 0x00007fcdbe460ea9 in monovm_execute_assembly (argc=0, argv=0x0, managedAssemblyPath=<optimized out>, exitCode=0x7fff2e4b7b9c) at /home/imran/ms/mono/dnr-90/runtime/src/mono/mono/mini/monovm.c:258
#20 0x000055e25b12bc07 in ExecuteManagedAssembly (currentExeAbsolutePath=0x55e25c35bc20 "./artifacts/tests/coreclr/Linux.x64.Release/Tests/Core_Root/corerun", clrFilesAbsolutePath=<optimized out>, managedAssemblyAbsolutePath=0x55e25c35bd30 "/home/imran/ms/mono/dnr-90/runtime/artifacts/tests/coreclr/Linux.x64.Release/baseservices/threading/DeadThreads/DeadThreads/DeadThreads.dll", managedAssemblyArgc=0, managedAssemblyArgv=0x0) at /home/imran/ms/mono/dnr-90/runtime/src/coreclr/src/hosts/unixcoreruncommon/coreruncommon.cpp:507
#21 0x000055e25b12a96e in main (argc=<optimized out>, argv=<optimized out>) at /home/imran/ms/mono/dnr-90/runtime/src/coreclr/src/hosts/unixcorerun/corerun.cpp:149
```

It's a little unusual that I haven't encountered this before; I'm investigating now.

**Review by naricc — APPROVED** [review]

File: eng/pipelines/mono/templates/xplat-pipeline-job.yml
Comment (safern) [inline eng/pipelines/mono/templates/xplat-pipeline-job.yml:51]: Is this now shared or why are we changing this?

Comment (imhameed) [reply eng/pipelines/mono/templates/xplat-pipeline-job.yml:51]: This was done to allow LLVM AOT Mono-specific jobs to use the one-and-only variant of CoreCLR that we build right now.

Support for CoreCLR `runtimeVariant`s was added in https://github.com/dotnet/runtime/pull/35841. But `runtimeVariant` isn't currently used at all when building CoreCLR. The choices I see right now are:
- remove `runtimeVariant` support entirely from eng/pipelines/coreclr/*.yml;
- build an `llvmaot` variant of CoreCLR, for later consumption by LLVM AOT Mono-specific jobs, that is identical to the non-variant CoreCLR; or
- explicitly depend on a shared non-variant CoreCLR.

Comment (safern) [reply eng/pipelines/mono/templates/xplat-pipeline-job.yml:51]: I think "explicitly depend on a shared non-variant CoreCLR." is the one that makes the most sense. I was mostly courious why was this change needed just to understand how the system worked.

It seems like we do have an extra `_` in between `CoreCLRProduct_` and `osGroup` though.

Comment (imhameed) [reply eng/pipelines/mono/templates/xplat-pipeline-job.yml:51]: > I think "explicitly depend on a shared non-variant CoreCLR." is the one that makes the most sense. I was mostly courious why was this change needed just to understand how the system worked.

Ah, got it.

> It seems like we do have an extra `_` in between `CoreCLRProduct_` and `osGroup` though.

Yeah. That's a consequence of how the product artifact name is concatenated: https://github.com/dotnet/runtime/pull/35841/files#diff-9d649abd6ea7b7db8a63e5dd8edfd14bR81

Comment (safern) [reply eng/pipelines/mono/templates/xplat-pipeline-job.yml:51]: I don't want to block this PR because of that since it seems like it is something we already heavily depend on and the change might be bigger that it seems.

File: eng/pipelines/runtime.yml
Comment (safern) [inline eng/pipelines/runtime.yml:498]: Why do we have to build mono AOT when libraries change?

Comment (imhameed) [reply eng/pipelines/runtime.yml]: Duplicated from the non-LLVM AOT template instantiation directly above. :)

Comment (SamMonoRT) [reply eng/pipelines/runtime.yml]: The goal was to run Mono lanes when there are Library changes, Mono changes and test changes. We have moved over to .NET BCL, so any change there can affect Mono, and we should test it out. (We just had a perf regression creep in due to a change in libraries a couple weeks ago !). I would recommend leaving the change as is, and see how badly this affects the queues/wait times for CI.

**Review by SamMonoRT — APPROVED** [review]
LGTM - please wait for Santi's approval prior to merging.

File: eng/pipelines/runtime.yml
Comment (safern) [reply eng/pipelines/runtime.yml]: It doesn't make sense to run runtime tests when libraries changes, libraries are not even used in that scope.

When libraries changes we run managed (libraries) tests using a mono runtime, that I agree 😄

**Review by safern — APPROVED** [review]

**Pull request merged**
